### PR TITLE
feat: add include_usage toggle and thinking pill for chat input

### PR DESCRIPTION
**Pull request by MaderHatt3r**
Add toggleable include_usage model option (Default/On/Off) that controls whether stream_options.include_usage is injected into streaming requests, allowing per-model compatibility control. Add a thinking toggle pill in the chat input area with a brain icon that lets users quickly enable or disable model reasoning (reasoning_effort) without switching models.

# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) to discuss your idea/fix with the community before creating a pull request, and describe your changes before submitting a pull request.

This is to ensure large feature PRs are discussed with the community first, before starting work on it. If the community does not want this feature or it is not relevant for Open WebUI as a project, it can be identified in the discussion before working on the feature and submitting the PR.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Verify that the pull request targets the `dev` branch. **PRs targeting `main` will be immediately closed.**
- [x] **Description:** Provide a concise description of the changes made in this pull request down below.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Add docs in [Open WebUI Docs Repository](https://github.com/open-webui/docs). Document user-facing behavior, environment variables, public APIs/interfaces, or deployment steps.
  - [x] Unlikely to need docs. Will defer to reviewer.
- [x] **Dependencies:** Are there any new or upgraded dependencies? If so, explain why, update the changelog/docs, and include any compatibility notes. Actually run the code/function that uses updated library to ensure it doesn't crash.
- [x] **Testing:** Perform manual tests to **verify the implemented fix/feature works as intended AND does not break any other functionality**. Include reproducible steps to demonstrate the issue before the fix. Test edge cases (URL encoding, HTML entities, types). Take this as an opportunity to **make screenshots of the feature/fix and include them in the PR description**.
- [x] **Agentic AI Code:** Confirm this Pull Request is **not written by any AI Agent** or has at least **gone through additional human review AND manual testing**. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
  - [x] Generative ai disclosure. Cursor was used extensively for codebase exploration, architectural understanding, and implementation assistance. All changes have been manually reviewed and tested.
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Design & Architecture:** Prefer smart defaults over adding new settings; use local state for ephemeral UI logic. Open a Discussion for major architectural or UX changes.
- [x] **Git Hygiene:** Keep PRs atomic (one logical change). Clean up commits and rebase on `dev` to ensure no unrelated commits (e.g. from `main`) are included. Push updates to the existing PR branch instead of closing and reopening.
- [x] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

Two quality-of-life additions for working with streaming LLM providers:

1. **Include Usage toggle** — A new model option (`include_usage`) that controls whether `stream_options: {include_usage: true}` is injected into streaming requests (needed for LM Stuido running models and possibly others). This enables token usage statistics (input/output tokens) from OpenAI-compatible providers that support it (including LM Studio, litellm, etc.). The toggle follows the same Default/On/Off pattern as the existing "Stream Chat Response" option, allowing per-model compatibility control for providers that don't support `stream_options`.

2. **Thinking toggle pill** — A brain icon pill button in the chat input area that lets users quickly toggle model reasoning (thinking) on or off without switching models or navigating to advanced settings. This is particularly useful for thinking-capable models (e.g., Qwen3, DeepSeek R1) where you want to disable reasoning for simple tasks to save time and tokens. The pill sends `reasoning_effort` as a parameter — `"off"` when disabled, or the user's configured value (defaulting to `"medium"`) when enabled. A tri-state toggle (Default/On/Off) is also available in the Integrations Menu.

### Added

- **`include_usage` model option** in Advanced Parameters (Default/On/Off toggle) — controls `stream_options.include_usage` injection for streaming requests (needed for LM Stuido running models)
- **Brain icon component** (`Brain.svelte`) — Lucide-style brain SVG for the thinking toggle
- **Thinking toggle pill** in the chat input area — purple when thinking is on, gray when off, hidden when default
- **Thinking toggle** in the Integrations Menu — tri-state (Default/On/Off) for discoverability
- Middleware injects `stream_options: {include_usage: true}` by default for all streaming requests to OpenAI-compatible providers (respects the `include_usage` toggle when explicitly set to Off)

### Changed

- **`MessageInput.svelte`** — Added brain pill display, `thinkingEnabled` state, and always-visible Integrations Menu button (so the thinking toggle is always discoverable)
- **`IntegrationsMenu.svelte`** — Added "Thinking" toggle entry with Brain icon
- **`Chat.svelte`** — Added `thinkingEnabled` state, bound to MessageInput and Placeholder, wired to `reasoning_effort` param in chat completion requests
- **`Placeholder.svelte`** — Added `thinkingEnabled` binding pass-through
- **`General.svelte`** — Added `include_usage` to saved params
- **`AdvancedParams.svelte`** — Added `include_usage` to default params and toggle UI
- **`middleware.py`** — Added `include_usage` to `open_webui_params` filter; conditional `stream_options` injection based on metadata param
- **`payload.py`** — Added `include_usage` to `open_webui_params` filter
- **`main.py`** — Added `include_usage` extraction from chat/model params and inclusion in metadata

### Deprecated

- N/A

### Removed

- N/A

### Fixed

- N/A

### Security

- N/A

### Breaking Changes

- None. Both features are additive. The `stream_options` injection defaults to the same behavior as before (always on). The thinking toggle defaults to null (no override), so existing behavior is unchanged.

---

### Additional Information

- **No new dependencies.** All changes use existing infrastructure — the `reasoning_effort` parameter was already supported in the OpenAI payload pipeline; this PR just adds a quick-access UI toggle for it.
- **`include_usage` compatibility note:** Some providers may not support `stream_options` and could return errors (haven't seen it cause issues as correctly implemented APIs should ignore extra fields, but better to be safe). The toggle allows users to disable it per-model for those cases. The default behavior (inject when not explicitly disabled) was chosen because most modern OpenAI-compatible providers support it.
- **Thinking toggle behavior:** The pill overrides any `reasoning_effort` value set in Advanced Parameters when explicitly toggled on or off. When set to Default (null), the Advanced Parameters value is used unchanged. This lets users set a preferred reasoning level in settings while still being able to quickly disable thinking from the chat input.
- Cursor was used extensively for codebase exploration, architectural understanding, and implementation assistance. All changes have been manually reviewed and tested.

### Screenshots or Videos

Doesn't show thinking icon indicator when `default`. Shows when `on` or `off`.

screenshot-2026-02-20 22_26_11-Arc
screenshot-2026-02-20 22_26_36-Arc
screenshot-2026-02-20 22_27_01-Arc

Note brain icon shows purple or grey only if NOT default value.

screenshot-2026-02-20 23_34_23-screenshot-2026-02-20 22_26_36-Arc (copy)

The stream_options --> include_usage addition makes my local LM Studio models properly return token usage.

image

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

> [!NOTE]
> Deleting the CLA section will lead to immediate closure of your PR and it will not be merged in.

### CODE REVIEW

File: src/lib/components/chat/MessageInput/IntegrationsMenu.svelte [hunk 1]
```diff
@@ -309,6 +311,40 @@
 							</button>
 						</Tooltip>
 					{/if}
+
+					<Tooltip content={$i18n.t('Toggle model thinking/reasoning')} placement="top-start">
```
Comment (MaderHatt3r) [inline src/lib/components/chat/MessageInput/IntegrationsMenu.svelte:315]: i18n used where appropriate. Some new.

Comment (MaderHatt3r) [reply src/lib/components/chat/MessageInput/IntegrationsMenu.svelte:315]: Do I need to `run npm run i18n:parse` myself or is that part of a workflow somewhere?

**Comment by MaderHatt3r** [comment]
> # ⚠️ Warning: Possible Non-Atomic / Scope Creep PR Detected
> Your PR was subjected to automated review by AI to determine if it could fall under Open WebUI's non-atomicity ruleset or scope creep.
> 
> **This PR appears to contain multiple unrelated changes that could be split into separate pull requests.**
> 
> ## 🔍 AI Analysis Summary
> **Primary Intent:** Add support for model thinking/reasoning toggle (reasoning_effort) with UI controls and include_usage stream option
> 
> **Secondary Changes Detected:**
> 
> * Add include_usage parameter for requesting token usage in streaming responses (backend)
> * Add UI toggle for thinking/reasoning control (frontend)
> * Add Brain icon component
> 
> 📝 Detailed Analysis and Full Report (click to expand)
> This PR bundles TWO distinct features: (1) The 'Thinking' toggle UI - a new user-facing feature to control model reasoning via reasoning_effort parameter with a Brain icon, toggle button in MessageInput, and menu option in IntegrationsMenu. (2) The include_usage backend feature - adding stream_options to request token usage statistics in streaming responses. These are architecturally independent: the thinking toggle is a new user feature affecting the frontend and params sent to models, while include_usage is a backend-only streaming enhancement. The 'Pre-factor' test confirms they could have been separate PRs: the include_usage streaming fix could have been merged independently before or after the thinking feature. Additionally, the Brain.svelte icon component could have been added as a separate maintenance task. The PR title would suggest these are compound intentions.
> 
> ## Why Atomic PRs With Narrow Scopes Matter
> Atomic PRs (single-purpose PRs) are:
> 
> * **Easier to review** - Reviewers can focus on one thing at a time
> * **Easier to test** - Each change can be verified independently
> * **Easier to revert** - If something breaks, we can revert just the problematic change
> * **Faster to merge** - Smaller, focused PRs get reviewed and merged quicker
> 
> ## What Makes a PR Atomic / Narrow in Scope?
> An atomic PR should contain **one semantic change**:
> 
> * ✅ Just one bug fix (even if it touches multiple files)
> * ✅ Just one feature (even if it requires changes across multiple files)
> * ✅ Just i18n/translation updates
> * ✅ Just documentation updates
> * ✅ Just refactoring of one specific thing
> * ✅ Just one performance improvement
> 
> ## What To Do
> _This is an automated analysis. If you believe this assessment is incorrect and your PR is actually atomic (all changes serve one unified purpose), please explain in a comment below._
> 
> Consider splitting this PR into separate, focused pull requests. Each PR should address **one specific thing**.
> 
> For example, if you have a bug fix and a new feature, submit them as two separate PRs.

It's not wrong but the overhead for creating these PRs is killing me and these are both very small compared to my others. Let me know if you'd like me to separate these two things.

1. stream_options --> include_usage option to indicate token usage should be returned
2. thinking toggle pill (brain icon)

**Comment by MaderHatt3r** [comment]
I saw that this was part of the PR pipeline/ workflow, so I ran formatters and translations. Translations seem like they could detract from reviewing a PR though. Let me know if I should revert translations for a simpler PR.

```yaml
- name: Format Frontend
  run: npm run format

- name: Run i18next
  run: npm run i18n:parse

- name: Check for Changes After Format
  run: git diff --exit-code    # <-- fails if ANY file changed
  ```

**Comment by Classic298** [comment]
General thinking Buttons are not wanted because different models require different thinking parameters.

Use a function to build your own thinking toggle in the Ui.

And for usage - that's already possible in the model editor?

Many models also don't support thinking or ONLY have thinking. How do you handle those cases?

Reasoning_effort is not supported by many providers.

Should be solved using your own function for your specific models. And for usage - you can already configure it in the model settings - and if you want a toggle for it - once again a filter with a button is the answer

Did you search for existing / previous PRs? This was proposed before (i think a few times even) and closed for this reason. Did you open a discussion? Something like this needs to be discussed first too.

**Comment by Classic298** [comment]
https://docs.openwebui.com/features/extensibility/plugin/functions/filter#--toggle-filter-example-adding-interactivity-and-icons-new-in-open-webui-0610

**Comment by Classic298** [comment]
To add some more context on the reasoning here:

The thinking/reasoning toggle is a good example of something that seems universal but isn't. reasoning_effort is an OpenAI-specific parameter - other providers handle reasoning control differently (or not at all), and some models are reasoning-only with no way to turn it off. A hardcoded toggle in the core chat UI would create an expectation that it works everywhere, when in reality it would silently do nothing or cause errors for most setups.

This is exactly the kind of thing the function/filter system was designed for. **You can build a toggle that targets your specific models and providers, with exactly the parameters they expect. The link I shared above has a working example with a UI button - it's essentially what you built here, just scoped to where it makes sense.**

On the include_usage side - the per-model toggle for that already exists in the model editor under Advanced Parameters. **Injecting stream_options into all requests by default would risk breaking providers that reject unknown fields.** The existing per-model approach is the safer granularity.

For future contributions, the PR checklist asks first-time contributors to open a Discussion before starting work on new features. That step would have surfaced these concerns early and saved you the implementation time. We'd encourage you to use that path going forward. It exists to help align effort with project direction before code gets written.

**Comment by MaderHatt3r** [comment]
I'm pretty sure I can post this as-is in the discussions and people would just say, great feature, thanks.

I didn't search existing PRs for this one because I wrote it for myself to use and it is such a simple, basic feature that almost everything else includes. I thought I'd go ahead and provide it back to the community. 

I agree, it could be confusing for anything that doesn't support it, but changing the settings via the model editor is tedious and looking for the setting in the panel. 

I can just keep this one on my personal working copy. Thanks for the feedback.

**Comment by Classic298** [comment]
@MaderHatt3r I agree editing all models is annoying. A universal-edit-all-models UI is, afaik, planned, which allows you to mass-edit all models to turn on the usage. I hope that helps

alternatively, as I do it right now, use a filter. The filter modifies all requests to always include all usage and with a single click, you can globally enable the filter. The filter has like 6 lines of code.

**Comment by MaderHatt3r** [comment]
It's not just that. I have to go edit the model for when I want it and edit it again for when I don't. It's intractable. The filter option is interesting, but since I already have this, I'll just keep it for myself. I like my cute little purple brain. lol. 

I suppose I could add the filter and release it as a plugin though so others can use it. I'll consider it. I like providing things I thought were useful back to the community. I always feel selfish to write something for myself and never share it.

**Comment by Classic298** [comment]
@MaderHatt3r you can add a togglable filter with a cute little purple brain as the icon as well just saying :)
togglable right from the chat

**Comment by Classic298** [comment]
@MaderHatt3r ABSOLUTELY! publish it on the community website and I'll be sure to upvote it and maybe if it gets traction it will be featured on the monthly newspaper
